sql/src/main/java/org/apache/druid/sql/calcite/schema/SystemSchema.java

```diff
@@ -0,0 +1,695 @@
+            {
+              try {
+                it.close();
+              }
+              catch (IOException e) {
+                throw new RuntimeException(e);
+              }
+            }
+          };
+        }
+      }
+
+      return new TasksEnumerable(getTasks(druidLeaderClient, jsonMapper, responseHandler));
+    }
+
+    private CloseableIterator<TaskStatusPlus> getAuthorizedTasks(JsonParserIterator<TaskStatusPlus> it, DataContext root)
+    {
+      final AuthenticationResult authenticationResult =
+          (AuthenticationResult) root.get(PlannerContext.DATA_CTX_AUTHENTICATION_RESULT);
+
+      Function<TaskStatusPlus, Iterable<ResourceAction>> raGenerator = task -> Collections.singletonList(
+          AuthorizationUtils.DATASOURCE_READ_RA_GENERATOR.apply(task.getDataSource()));
+
+      final Iterable<TaskStatusPlus> authorizedTasks = AuthorizationUtils.filterAuthorizedResources(
+          authenticationResult, () -> it, raGenerator, authorizerMapper);
+
+      return wrap(authorizedTasks.iterator(), it);
+    }
+
+  }
+
+  //Note that overlord must be up to get tasks
+  private static JsonParserIterator<TaskStatusPlus> getTasks(
+      DruidLeaderClient indexingServiceClient,
+      ObjectMapper jsonMapper,
+      BytesAccumulatingResponseHandler responseHandler
+  )
+  {
+
+    Request request;
+    try {
+      request = indexingServiceClient.makeRequest(
+          HttpMethod.GET,
+          StringUtils.format("/druid/indexer/v1/tasks")
+      );
+    }
+    catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+    ListenableFuture<InputStream> future = indexingServiceClient.goAsync(
+        request,
+        responseHandler
+    );
+    try {
+      future.get();
+    }
+    catch (InterruptedException | ExecutionException e) {
+      throw new RuntimeException(e);
+    }
+    if (responseHandler.getStatus() != HttpServletResponse.SC_OK) {
+      throw new ISE(
+          "Error while fetching tasks status[%s] description[%s]",
+          responseHandler.status,
+          responseHandler.description
+      );
+    }
+    final JavaType typeRef = jsonMapper.getTypeFactory().constructType(new TypeReference<TaskStatusPlus>()
+    {
+    });
+    return new JsonParserIterator<>(
+        typeRef,
+        future,
+        request.getUrl().toString(),
+        null,
+        request.getUrl().getHost(),
+        jsonMapper
+    );
+  }
+
+  private static <T> CloseableIterator<T> wrap(Iterator<T> iterator, JsonParserIterator<T> it)
+  {
+    return new CloseableIterator<T>()
+    {
+      @Override
+      public boolean hasNext()
+      {
+        return iterator.hasNext();
+      }
+
+      @Override
+      public T next()
+      {
+        return iterator.next();
+      }
+
+      @Override
+      public void close() throws IOException
+      {
+        it.close();
+      }
+    };
+  }
+
+  static class BytesAccumulatingResponseHandler extends InputStreamResponseHandler
```

Comment:
There are two more same classes. Please extract them to the public class and reuse it.

Reply:
okay, created a common public class.